### PR TITLE
Adds ListReleaseContents

### PR DESCRIPTION
**Pull request by JosephSalisbury**
Towards https://github.com/giantswarm/giantswarm/issues/4326

### CODE REVIEW

**Comment by JosephSalisbury** [comment]
@giantswarm/sig-content ignore, misclick

File: helmclient.go [hunk 1]
```diff
@@ -425,6 +413,49 @@ func (c *Client) InstallFromTarball(path, ns string, options ...helmclient.Insta
 	return nil
 }
 
+func (c *Client) ListReleaseContents() ([]*ReleaseContent, error) {
```
Comment (kopiczko) [inline helmclient.go]: Please pass `ctx context.Context` as an argument. See https://github.com/giantswarm/helmclient/pull/58.

File: helmclient.go [hunk 2]
```diff
@@ -425,6 +413,49 @@ func (c *Client) InstallFromTarball(path, ns string, options ...helmclient.Insta
...
+			if err != nil {
+				return microerror.Mask(err)
+			}
+
+			return nil
+		}
+		b := backoff.NewExponential(2*time.Minute, 60*time.Second)
+		n := func(err error, delay time.Duration) {
+			c.logger.Log("level", "debug", "message", "failed listing release contents")
+		}
```
Comment (kopiczko) [inline helmclient.go]: Backoff notifications and intervals code will also change in https://github.com/giantswarm/helmclient/pull/58. Let's make it right from the beggining.

File: helmclient.go [hunk 3]
```diff
@@ -605,3 +636,26 @@ func getPodName(client kubernetes.Interface, labelSelector, namespace string) (s
 
 	return pod.Name, nil
 }
+
+func releaseToReleaseContent(release *hapirelease.Release) (*ReleaseContent, error) {
+	// If parameterizable values were passed at release creation time, raw values
+	// are returned by the Tiller API and we convert these to a map. First we need
+	// to check if there are values actually passed.
+	var values chartutil.Values
+	var err error
```
Comment (kopiczko) [inline helmclient.go]:
```suggestion
```

File: helmclient.go [hunk 4]
```diff
@@ -605,3 +636,26 @@ func getPodName(client kubernetes.Interface, labelSelector, namespace string) (s
 
 	return pod.Name, nil
 }
+
+func releaseToReleaseContent(release *hapirelease.Release) (*ReleaseContent, error) {
```
Comment (kopiczko) [inline helmclient.go]:
```suggestion
func releaseToReleaseContent(release *hapirelease.Release) (*ReleaseContent, error) {
	var err error
	
```

File: helmclient.go [hunk 5]
```diff
@@ -425,6 +413,49 @@ func (c *Client) InstallFromTarball(path, ns string, options ...helmclient.Insta
...
+		o := func() error {
+			t, err := c.newTunnel()
+			if IsTillerNotFound(err) {
+				return backoff.Permanent(microerror.Mask(err))
+			} else if err != nil {
+				return microerror.Mask(err)
+			}
+			defer c.closeTunnel(t)
+
+			resp, err = c.newHelmClientFromTunnel(t).ListReleases()
```
Comment (kopiczko) [inline helmclient.go]: I don't know the api but maybe it requires paging.

Comment (JosephSalisbury) [reply helmclient.go]: Possibly. We have < 10 releases, so I'm not too worried about it.

Comment (kopiczko) [reply helmclient.go]: I am. I can recall problems with paging ELBs. Recently with paging VMSS instances..

```
// ListReleasesResponse is a list of releases.
type ListReleasesResponse struct {
	// Count is the expected total number of releases to be returned.
	Count int64 `protobuf:"varint,1,opt,name=count" json:"count,omitempty"`
	// Next is the name of the next release. If this is other than an empty
	// string, it means there are more results.
	Next string `protobuf:"bytes,2,opt,name=next" json:"next,omitempty"`
	// Total is the total number of queryable releases.
	Total int64 `protobuf:"varint,3,opt,name=total" json:"total,omitempty"`
	// Releases is the list of found release objects.
	Releases []*hapi_release5.Release `protobuf:"bytes,4,rep,name=releases" json:"releases,omitempty"`
}
```

Check `Next`.

File: helmclient.go (hunk 1 above)
Comment (JosephSalisbury) [reply helmclient.go]: Done

File: helmclient.go (hunk 2 above)
Comment (JosephSalisbury) [reply helmclient.go]: Done

File: helmclient.go (hunk 3 above)
Comment (JosephSalisbury) [reply helmclient.go]: Done, just not via GitHub :D

File: helmclient.go (hunk 4 above)
Comment (JosephSalisbury) [reply helmclient.go]: Done, just not via GitHub again

File: helmclient.go (hunk 5 above)
Comment (JosephSalisbury) [reply helmclient.go]: Then I shall worry about it

File: helmclient.go [hunk 6]
```diff
@@ -425,6 +413,47 @@ func (c *Client) InstallFromTarball(path, ns string, options ...helmclient.Insta
...
+			if err != nil {
+				return microerror.Mask(err)
+			}
+
+			return nil
+		}
+		b := backoff.NewExponential(2*time.Minute, 60*time.Second)
```
Comment (kopiczko) [inline helmclient.go]:
```suggestion
		b := backoff.NewExponential(backoff.ShortMaxWait, backoff.ShortMaxInterval)
```

Comment (JosephSalisbury) [reply helmclient.go]: Pushed. Swear I had this earlier :D

File: helmclient.go (hunk 5 above)
Comment (JosephSalisbury) [reply helmclient.go]: Easy man, I can not be lazy heh

**Comment by JosephSalisbury** [comment]
@kopiczko mind having another look? i think we're good to go here

File: helmclient.go (hunk 5 above)
Comment (JosephSalisbury) [reply helmclient.go]: Done

File: helmclient.go [hunk 7]
```diff
@@ -425,6 +427,69 @@ func (c *Client) InstallFromTarball(path, ns string, options ...helmclient.Insta
...
+					return microerror.Mask(err)
+				}
+
+				releases = append(releases, resp.GetReleases()...)
+
+				if resp.GetNext() == "" {
+					moreReleases = false
+				} else {
+					next = resp.GetNext()
+				}
```
Comment (kopiczko) [inline helmclient.go]: Please just replace with `for {`. There is no need to maintain another variable.

```go
next = resp.GetNext()
if next == "" {
	break
}
```

Comment (JosephSalisbury) [reply helmclient.go]: Done

**Comment by JosephSalisbury** [comment]
@kopiczko ptala

File: spec.go [hunk 8]
```diff
@@ -35,8 +35,10 @@ type Interface interface {
 	// The releaseName is the name of the Helm Release that is set when the Helm
 	// Chart is installed.
 	GetReleaseHistory(ctx context.Context, releaseName string) (*ReleaseHistory, error)
-	// InstallFromTarball installs a Helm Chart packaged in the given tarball.
+	// InstallReleaseFromTarball installs a Helm Chart packaged in the given tarball.
```
Comment (JosephSalisbury) [inline spec.go]: Ninja fix

**Pull request merged**